backbone.js

```diff
@@ -1088,8 +1088,8 @@
     },
 
     // Define how to uniquely identify models in the collection.
-    modelId: function(attrs) {
-      return attrs[this.model.prototype.idAttribute || 'id'];
+    modelId: function(attrs, model) {
+      return attrs[(model && model.idAttribute) || this.model.prototype.idAttribute || 'id'];
```

Comment:
I found a major hole in this PR because of this comment, thank you! 

The _byId cache stores models in the collection by both `cid` and `id`, so by modifying `modelId` to return the id of polymorphic elements, we run the risk of cache conflicts. Ex: A mixed entities collection containing both companies and users where there is a company with `companyId: 1` and a user with `userId: 1`. This could cause issues.

To fix this, the _byId cache would need to track models internally by the unique combination of idAttribute and id:  `_byId[model.idAttribute + this.modelId(model.attributes, model.idAttribute)] = model`.

However, then `Collection.get` will no longer be able to find models by `id` at all any more.

We'd need to tweak `Collection.get` to do something like this:
```
    // Get a model from the set by cid, id, model object with id or cid
    // properties, or an attributes object that is transformed through modelId.
    // When fetching by id, use the optional idAttribute if provided, otherwise the 
    // provided model's idAttribute, or the collection's template model's idAttribute
     get: function(obj, idAttribute) {
      if (obj == null) return void 0;
      idAttribute = idAttribute || obj.idAttribute  || this.model.prototype.idAttribute;
      return this._byId[obj] ||
        this._byId(idAttribute + obj) ||
        this._byId[idAttribute + this.modelId(obj.attributes || obj, idAttribute)] ||
        obj.cid && this._byId[obj.cid];
    },
```
This just got a little more complicated, but what do you think? Messing with how models are stored in `_byId` is a major change...